src/core/lib/transport/promise_endpoint.h

```diff
@@ -0,0 +1,315 @@
+      grpc_core::MutexLock lock(&write_mutex_);
+      if (!write_result_.has_value()) {
+        write_waker_ = grpc_core::Activity::current()->MakeNonOwningWaker();
+        return grpc_core::Pending();
+      } else {
+        const auto ret = *write_result_;
+        write_result_.reset();
+        return ret;
+      }
+    };
+  }
+
+  auto Read(size_t num_bytes) {
+    grpc_core::MutexLock lock(&read_mutex_);
+
+    /// Previous read result has not been polled.
+    GPR_ASSERT(!read_result_.has_value());
+
+    /// Should not have pending reads
+    GPR_ASSERT(pending_read_buffer_.Count() == 0u);
+
+    if (read_buffer_.Count() < num_bytes) {
+      const std::function<void(absl::Status)> read_callback =
+          [this, num_bytes, &read_callback](absl::Status status) {
+            if (!status.ok()) {
+              pending_read_buffer_.Clear();
+              temporary_read_buffer_.Clear();
+
+              grpc_core::MutexLock lock(&read_mutex_);
+              read_result_ = status;
+              read_waker_.Wakeup();
+            } else {
+              while (temporary_read_buffer_.Count() > 0u) {
+                pending_read_buffer_.Append(temporary_read_buffer_.TakeFirst());
+              }
+
+              if (read_buffer_.Count() + pending_read_buffer_.Count() <
+                  num_bytes) {
+                /// A further read is needed.
+                endpoint_->Read(read_callback, &temporary_read_buffer_,
+                                nullptr /* uses default arguments */);
+              } else {
+                while (pending_read_buffer_.Count() > 0u) {
+                  grpc_slice_buffer_add(
+                      read_buffer_.c_slice_buffer(),
+                      pending_read_buffer_.TakeFirst().c_slice());
+                }
+
+                grpc_core::MutexLock lock(&read_mutex_);
+                read_result_ = status;
+                read_waker_.Wakeup();
+              }
+            }
+          };
+      temporary_read_buffer_.Clear();  /// may be redundant
+      endpoint_->Read(read_callback, &temporary_read_buffer_,
+                      nullptr /* uses default arguments */);
+    } else {
+      read_result_ = absl::OkStatus();
+    }
+
+    return [this, num_bytes]()
+               -> grpc_core::Poll<absl::StatusOr<grpc_core::SliceBuffer>> {
+      grpc_core::MutexLock lock(&read_mutex_);
+      if (!read_result_.has_value()) {
+        read_waker_ = grpc_core::Activity::current()->MakeNonOwningWaker();
+        return grpc_core::Pending();
+      } else if (!read_result_->ok()) {
+        const absl::Status ret = *read_result_;
+        read_result_.reset();
+
+        return ret;
+      } else {
+        grpc_core::SliceBuffer ret;
+        grpc_slice_buffer_move_first(read_buffer_.c_slice_buffer(), num_bytes,
+                                     ret.c_slice_buffer());
+
+        read_result_.reset();
+        return ret;
+      }
+    };
+  }
+
+  auto ReadSlice(size_t num_bytes) {
```

Comment:
Is it possible to re-use some code from Read() implementation inside this function ?

Reply:
I have thought about that. But I think there is some optimization we can do in this special case, so I chose not to.

Reply:
Worth considering whether the aggregation code can be moved into `SliceBuffer` however (although I'll not insist that happens)